### PR TITLE
Add persistent IDs to partition fields

### PR DESCRIPTION
**Pull request by jun-he**
For issue https://github.com/apache/incubator-iceberg/issues/280.

I took a different approach from the previous WIP PR (https://github.com/apache/incubator-iceberg/pull/499).
Instead of adding additional states and logics into `TableMetadata`, I created `PartitionSpecUpdate` to handle that. 
Also, it is unnecessary to keep `lastAssignedPartitonFieldId` in `TableMetadata` as it can be lazily derived from all specs (previous specs are still there) stored in `TableMetadata`.

### CODE REVIEW

**Comment by jun-he** [comment]
@rdblue can you please take a look? Thanks.

**Comment by jun-he** [comment]
Hi @rdblue, can you please take a look? Thanks.

File: core/src/main/java/org/apache/iceberg/PartitionSpecParser.java [hunk 1]
```diff
@@ -147,7 +156,12 @@ private static void buildFromJsonFields(PartitionSpec.Builder builder, JsonNode
       String transform = JsonUtil.getString(TRANSFORM, element);
       int sourceId = JsonUtil.getInt(SOURCE_ID, element);
 
-      builder.add(sourceId, name, transform);
+      // partition field ids are missing in old PartitionSpec, they always auto-increment from PARTITION_DATA_ID_START
+      if (!hasFieldId) {
```
Comment (chenjunjiedada) [inline core/src/main/java/org/apache/iceberg/PartitionSpecParser.java:160]: How about the forward compatibility? Is it possible that an old reader reads the new spec? then it still parses the new spec field id start from 1000?

File: api/src/main/java/org/apache/iceberg/PartitionField.java [hunk 2]
```diff
@@ -28,11 +28,13 @@
...
 public class PartitionField implements Serializable {
   private final int sourceId;
+  private final int fieldId;
   private final String name;
   private final Transform<?, ?> transform;
 
-  PartitionField(int sourceId, String name, Transform<?, ?> transform) {
+  PartitionField(int sourceId, int fieldId, String name, Transform<?, ?> transform) {
     this.sourceId = sourceId;
+    this.fieldId = fieldId;
```
[the remark this answers is not in the document]
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/PartitionField.java:37]: I don't think so. That is a convention that we use, but not strictly required by the spec.

File: api/src/main/java/org/apache/iceberg/PartitionField.java [hunk 3]
```diff
@@ -60,7 +69,7 @@ public String name() {
 
   @Override
   public String toString() {
-    return name + ": " + transform + "(" + sourceId + ")";
+    return name + " (" + fieldId + "): " + transform + "(" + sourceId + ")";
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/PartitionField.java:72]: Nit: schemas use the format `1000: ts_day`. That would be a bit cleaner than two sets of parens.

File: api/src/main/java/org/apache/iceberg/PartitionSpec.java [hunk 4]
```diff
@@ -58,14 +58,16 @@
   private transient volatile ListMultimap<Integer, PartitionField> fieldsBySourceId = null;
   private transient volatile Class<?>[] lazyJavaClasses = null;
   private transient volatile List<PartitionField> fieldList = null;
+  private final int lastAssignedFieldId;
```
[the remark this answers is not in the document]
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/PartitionSpec.java:61]: This is tracking something different. Here, this is the highest ID assigned to any partition, so that the next ID assigned will be unique.

File: api/src/main/java/org/apache/iceberg/PartitionSpec.java [hunk 5]
```diff
@@ -307,11 +312,16 @@ public static Builder builderFor(Schema schema) {
     private final Set<String> partitionNames = Sets.newHashSet();
     private Map<Integer, PartitionField> timeFields = Maps.newHashMap();
     private int specId = 0;
+    private int lastAssignedFieldId = PARTITION_DATA_ID_START - 1;
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/PartitionSpec.java:315]: We tend to prefer using `AtomicInteger` instead of an `int` field. That way it is more clear because the calls are `incrementAndGet` or `getAndIncrement`. That would also avoid the need to start at `PARTITION_DATA_ID_START - 1`. You could use the constant to initialize and assign using `getAndIncrement`.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 6]
```diff
@@ -350,7 +354,8 @@ public TableMetadata updateSchema(Schema newSchema, int newLastColumnId) {
         currentSnapshotId, snapshots, snapshotLog, addPreviousFile(file, lastUpdatedMillis));
   }
 
-  public TableMetadata updatePartitionSpec(PartitionSpec newPartitionSpec) {
+  // newPartitionSpec's partition field IDs should have already been refreshed
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:357]: Spec field IDs don't need to be refreshed. `freshSpec` is used for two things:
1. If schema IDs are reassigned, then it rebuilds the spec to with the new schema
2. It sets the spec's ID to the one assigned in this method

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 7]
```diff
@@ -350,7 +354,8 @@ public TableMetadata updateSchema(Schema newSchema, int newLastColumnId) {
...
+  TableMetadata updatePartitionSpec(PartitionSpec newPartitionSpec) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:358]: These should still be public like the other methods. Instead of making them private, they should sanity check the spec:
* For a v1 table, field IDs should be assigned from 1000 and should have no gaps
* For a v2 table or newer, all fields should have IDs

File: core/src/main/java/org/apache/iceberg/PartitionSpecParser.java (hunk 1 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/PartitionSpecParser.java:160]: Instead of making these changes here, I think that this should be verified in `TableMetadata`. That would accomplish the same thing, but make it easier to check the table version.

File: api/src/main/java/org/apache/iceberg/UpdatePartitionSpec.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+import org.apache.iceberg.exceptions.CommitFailedException;
+
+/**
+ * API for partition spec evolution.
+ * <p>
+ * When committing, these changes will be applied to the current table metadata. Commit conflicts
+ * will not be resolved and will result in a {@link CommitFailedException}.
+ */
+public interface UpdatePartitionSpec extends PendingUpdate<PartitionSpec> {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/UpdatePartitionSpec.java:30]: Can we remove the new interface and methods from this PR?

I don't think this is needed for this PR, and I'd like to minimize the number of changes. In addition, I don't think we want to move to a model where users create a new spec and apply it to a table. I think we instead want to evolve the partition spec of a table. So this API will probably be different when we release that feature.

**Comment by rdblue** [comment]
Thanks @jun-he, this is great progress on partition field IDs! I think I understand how compatibility will work with field IDs and I tried to add that context to my comments in this PR. If it isn't clear, please let me know what doesn't make sense.

File: core/src/test/java/org/apache/iceberg/TestMergeAppend.java [hunk 9]
```diff
@@ -395,6 +396,47 @@ public void testChangedPartitionSpec() {
         initialManifest, pending.manifests().get(1));
   }
 
+  @Test
+  public void testUpdatePartitionSpecFieldIds() {
+    TableMetadata base = readMetadata();
+
+    // build the new spec using the table's schema, which uses fresh IDs
+    PartitionSpec newSpec = PartitionSpec.builderFor(base.schema())
```
[the remark this answers is not in the document]
Comment (jun-he) [reply core/src/test/java/org/apache/iceberg/TestMergeAppend.java:404]: Done. Thanks.

File: api/src/main/java/org/apache/iceberg/PartitionSpec.java (hunk 5 above)
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/PartitionSpec.java:315]: I think we still have to use `PARTITION_DATA_ID_START - 1` to handle the case that there is no assigned field Id. Or we can update this field as `nextFieldId` and then set it from `PARTITION_DATA_ID_START`.

File: api/src/main/java/org/apache/iceberg/PartitionSpec.java [hunk 10]
```diff
@@ -437,24 +447,30 @@ public Builder truncate(String sourceName, int width, String targetName) {
...
       return this;
     }
 
     public Builder truncate(String sourceName, int width) {
       return truncate(sourceName, width, sourceName + "_trunc");
     }
 
-    Builder add(int sourceId, String name, String transform) {
+    // add a partition field with an auto-increment partition field id starting from PARTITION_DATA_ID_START
+    Builder append(int sourceId, String name, String transform) {
```
[the remark this answers is not in the document]
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/PartitionSpec.java:459]: Sure, change it back to `add`.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 6 above)
Comment (jun-he) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:357]: This comment is about the input parameter `PartitionSpec newPartitionSpec`, whose field ids has to be processed/refreshed so that `freshSpec` does not need to take care of that extra work.

File: api/src/main/java/org/apache/iceberg/PartitionField.java (hunk 3 above)
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/PartitionField.java:72]: 👌

File: api/src/main/java/org/apache/iceberg/UpdatePartitionSpec.java (hunk 8 above)
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/UpdatePartitionSpec.java:30]: I agree that it is better to move the update to a new PR, which addresses #281. 
The main reason I put it here is to have additional unit tests to make sure it works as expected.

Additionally, I think we may support the table partition spec evolution in two ways
```
table.updatePartitionSpec()
  .update(spec)
  .commit();
```
```
table.updatePartitionSpec().newSpec(schema)
  .identity(...)
  .bucket(...)
  ...
  .commit();
```
The first approach may be used if clients want to define a spec and manage it by their codes, e.g. use defined spec object in multiple places.

File: core/src/main/java/org/apache/iceberg/PartitionSpecParser.java [hunk 11]
```diff
@@ -138,6 +140,13 @@ private static void buildFromJsonFields(PartitionSpec.Builder builder, JsonNode
         "Cannot parse partition spec fields, not an array: %s", json);
 
     Iterator<JsonNode> elements = json.elements();
+
+    boolean hasFieldId = false;
+    if (elements.hasNext() && elements.next().has(FIELD_ID)) {
+      hasFieldId = true;
+    }
+    elements = json.elements();
```
[the remark this answers is not in the document]
Comment (jun-he) [reply core/src/main/java/org/apache/iceberg/PartitionSpecParser.java:148]: 👌

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 7 above)
Comment (jun-he) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:358]: As this check and field ID change will be in the same PR,  do we still need to check if all fields should have IDs in `newPartitionSpec`?  When the check is called, the PartitionSpec must have a field ID, right?

**Comment by jun-he** [comment]
I will move `UpdatePartitionSpec` and all related unit tests to another PR.

File: api/src/main/java/org/apache/iceberg/Transaction.java [hunk 12]
```diff
@@ -40,6 +40,14 @@
    */
   UpdateSchema updateSchema();
 
+
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/Transaction.java:43]: Nit: unnecessary newline.

File: api/src/main/java/org/apache/iceberg/UpdatePartitionSpec.java (hunk 8 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/UpdatePartitionSpec.java:30]: For tests, you can use `TableMetadata` and `commit` directly:

```java
TableOperations ops = table.operations();
TableMetadata base = ops.current()
TableMetadata updated = base.updatePartitionSpec(newSpec);
ops.commit(base, updated);
```

File: core/src/main/java/org/apache/iceberg/PartitionSpecParser.java [hunk 13]
```diff
@@ -138,6 +140,7 @@ private static void buildFromJsonFields(PartitionSpec.Builder builder, JsonNode
         "Cannot parse partition spec fields, not an array: %s", json);
 
     Iterator<JsonNode> elements = json.elements();
+    int fieldIdCnt = 0;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/PartitionSpecParser.java:143]: Nit: we like to avoid abbreviations that aren't readable. Saving a few chars isn't worth trading readability.

File: core/src/main/java/org/apache/iceberg/PartitionSpecParser.java [hunk 14]
```diff
@@ -147,7 +150,17 @@ private static void buildFromJsonFields(PartitionSpec.Builder builder, JsonNode
...
+      if (element.has(FIELD_ID)) {
+        builder.add(sourceId, JsonUtil.getInt(FIELD_ID, element), name, transform);
+        fieldIdCnt++;
+      }  else {
+        builder.add(sourceId, name, transform);
+      }
     }
+
+    Preconditions.checkArgument(fieldIdCnt == 0 || fieldIdCnt == json.size(),
+        "Parsed an invalid PartitionSpec Json. Some of its fields (%s/%s) miss field IDs.",
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/PartitionSpecParser.java:163]: How about "Cannot parse spec with missing field IDs: %s missing of %s fields"?

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 15]
```diff
@@ -208,8 +210,9 @@ public String toString() {
     this.lastUpdatedMillis = lastUpdatedMillis;
     this.lastColumnId = lastColumnId;
     this.schema = schema;
-    this.specs = specs;
     this.defaultSpecId = defaultSpecId;
+    this.specs = specs;
+
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:215]: Can you revert these unnecessary changes? The order of assignments changed and there's an extra newline. These non-functional changes can cause git conflicts, so we try to avoid them.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 16]
```diff
@@ -487,8 +491,9 @@ public TableMetadata removeSnapshotLogEntries(Set<Long> snapshotIds) {
         currentSnapshotId, snapshots, newSnapshotLog, addPreviousFile(file, lastUpdatedMillis));
   }
 
-  public TableMetadata buildReplacement(Schema updatedSchema, PartitionSpec updatedPartitionSpec,
-                                        Map<String, String> updatedProperties) {
+  // updatedPartitionSpec's partition field IDs should have already been refreshed
+  TableMetadata buildReplacement(Schema updatedSchema, PartitionSpec updatedPartitionSpec,
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:495]: Why is this no longer public?

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 17]
```diff
@@ -352,6 +355,7 @@ public TableMetadata updateSchema(Schema newSchema, int newLastColumnId) {
         currentSnapshotId, snapshots, snapshotLog, addPreviousFile(file, lastUpdatedMillis));
   }
 
+  // Input newPartitionSpec's partition field IDs should have already been recomputed
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:358]: #903 added `formatVersion` to metadata, so now we can add a validation here when the table is v1 that the spec IDs are assigned as expected:

```
ValidationException.check(formatVersion > 1 || hasSequentialIds(spec),
    "Spec does not use sequential IDs that are required in v1: %s", spec);
```

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 18]
```diff
@@ -487,8 +491,9 @@ public TableMetadata removeSnapshotLogEntries(Set<Long> snapshotIds) {
         currentSnapshotId, snapshots, newSnapshotLog, addPreviousFile(file, lastUpdatedMillis));
   }
 
-  public TableMetadata buildReplacement(Schema updatedSchema, PartitionSpec updatedPartitionSpec,
-                                        Map<String, String> updatedProperties) {
+  // updatedPartitionSpec's partition field IDs should have already been refreshed
+  TableMetadata buildReplacement(Schema updatedSchema, PartitionSpec updatedPartitionSpec,
+                                 Map<String, String> updatedProperties) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:496]: We should add a validation for format v1 here as well.

File: core/src/test/java/org/apache/iceberg/TestMergeAppend.java [hunk 19]
```diff
@@ -373,10 +374,10 @@ public void testChangedPartitionSpec() {
         .bucket("id", 4)
         .build();
 
-    // commit the new partition spec to the table manually
-    table.ops().commit(base, base.updatePartitionSpec(newSpec));
+    // commit the new partition spec to the table
+    table.updatePartitionSpec().update(newSpec).commit();
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestMergeAppend.java:378]: As I mentioned earlier, I think this should continue using the `updatePartitionSpec` method so that we can add a public API for this in a separate commit.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 7 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java]: You're right. We just need to check that ids are sequential and starting at 1,000 if the table version is v1.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 16 above)
Comment (jun-he) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:495]: @rdblue  thanks for the comment.
I am thinking to expose those functionalities (`PartitionSpecUpdate` or `buildReplacement`) over `Table` and `UpdatePartitionSpec` public interface. So we can limit their access to be package/default level.

I will change them back to public in the PR and we may limit their access in the partition spec evolution PR.

File: api/src/main/java/org/apache/iceberg/UpdatePartitionSpec.java (hunk 8 above)
Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/UpdatePartitionSpec.java:30]: 👌

File: core/src/test/java/org/apache/iceberg/TestMergeAppend.java (hunk 19 above)
Comment (jun-he) [reply core/src/test/java/org/apache/iceberg/TestMergeAppend.java:378]: 👌

**Comment by jun-he** [comment]
@chenjunjiedada FYI, I refactored the code and the partition spec evolution changes and related tests are now in https://github.com/apache/incubator-iceberg/pull/922. Thanks.

File: core/src/test/java/org/apache/iceberg/TestMergeAppend.java [hunk 20]
```diff
@@ -648,4 +649,98 @@ public void testInvalidAppendManifest() throws IOException {
             .appendManifest(manifestWithDeletedFiles)
             .commit());
   }
+
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestMergeAppend.java:652]: Nit: unnecessary newline.

**Comment by rdblue** [comment]
Awesome work, @jun-he! I'm merging this.

File: core/src/test/java/org/apache/iceberg/TestMergeAppend.java (hunk 20 above)
Comment (jun-he) [reply core/src/test/java/org/apache/iceberg/TestMergeAppend.java:652]: Thanks. Will remove it in https://github.com/apache/incubator-iceberg/pull/922.
